### ISSUE TITLE
[$1000] Android: Cursor moves around wildly after pasting emoji reported by @adeel0202 

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#12578]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Actions performed:

1. Go to any chat
2. In the composer, type an emoji alias with a space before ending colon e.g. :wave :
3. Set position of the cursor before the ending colon
4. Press clear key from keyboard to remove the space
5. Press clear key from keyboard to remove the emoji

(Note: it doesn't crash everytime so you might have to try multiple times with different emojis)

## Expected Result:
Emoji is removed

## Actual Result:
App crashes

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Android

**Version Number:** 1.2.25-0

**Reproducible in staging?:** y

**Reproducible in production?:** y

**Email or phone of affected tester (no customers):** 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/200717691-63626773-f878-4cd5-acb7-bc6823f30158.mp4

https://user-images.githubusercontent.com/43996225/200717706-1cd5716d-014c-4861-abdb-ad048ffbdb1b.mp4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** @adeel0202 

**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1667905856603069

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

## Upwork Automation - Do Not Edit
- Upwork Job URL: https://www.upwork.com/jobs/~011341b7bfc98aa844
- Upwork Job ID: 1595148323923935232

**Comment by cristipaval** [comment]
I think this is a good starting point to contribute on NewDot

**Comment by JmillsExpensify** [comment]
Assigning myself as part of the broader emoji initiative.

**Comment by JmillsExpensify** [comment]
Now that I've added myself haha, let me test this issue and keep it moving forward.

**Comment by JmillsExpensify** [comment]
Well shoot, as soon as I say that, I look at the issue and I can't test this because it's an Android issue and I have no dev/real device. That said, this is a crash, so we should have a corresponding issue automatically created via Firebase. I'm going to search the repo now to see what's up.

**Comment by JmillsExpensify** [comment]
In terms of likely candidates, I think this crash most closely mimics the reproduction steps. @cristipaval mind giving this a closer look? This is the [search query](https://github.com/Expensify/App/issues?q=is%3Aissue+investigate+crash+in%3Atitle+) for the automatically created Firebase issues, and then you can toggle on closed to find this crash in particular.

**Comment by cristipaval** [comment]
I tested on 2 Samsung phones and on the emulator. I tested on dev with the latest main. 
The App doesn't not crash but the cursor moves weird.

https://user-images.githubusercontent.com/18078393/201210919-e4211f8b-1ca1-43af-8a8d-04a8129a0cb3.mp4

**Comment by JmillsExpensify** [comment]
That's super helpful! Ok we actually have several cursor issues in the /App repo, so I'll circle back tomorrow and see if this already exists. If it does, as I vaguely remember seeing something somewhere, then I will close. If not, then I'll update this issue to focus away from the crash and more so the cursory.

**Comment by JmillsExpensify** [comment]
Circling back on this, I wasn't able to find any cursor issues that approximated this one. There were actually only three in the repo. I'm going to update the title accordingly.

**Comment by JmillsExpensify** [comment]
Now that I'm acting on the above, @cristipaval do you mind sharing the reproduction steps from your recording. It's a bit hard to follow. For instance, before the cursor starts moving around wildly, did you paste an emoji?

**Comment by JmillsExpensify** [comment]
Then one more thing until we open this up for proposals. @kavimuru you still able to produce this Android cash?

**Comment by JmillsExpensify** [comment]
@adeel0202 Thanks for reporting. Can you clarify if this is in the simulator? Or if it's on an actual device, the model and Android version?

**Comment by JmillsExpensify** [comment]
Ok that's helpful, though it also means that the next steps are less clear, as at the moment we are not focusing on mobile crashes that occur less than ten times. It should also be captured though in [this list](https://github.com/Expensify/App/issues?q=is%3Aissue+investigate+crash+in%3Atitle+), which is the sum of all crashes captured by Firebase.

**Comment by JmillsExpensify** [comment]
@adeel0202 Thanks for your patience. Since my last comment, we've gotten a bit more crisp on our approach to crashes. We're going to address any crash that is reproducible, just like any other bug. Can you please update the title, reproduction steps and the specific information about your device above? Then we can test out seeing if anyone picks up this OnePlus crash.

**Comment by cristipaval** [comment]
Hey @adeel0202 !

Could you also go to your keyboard settings on your phone and send me a screenshot? It would also help to see if your phone uses a specific keyboard.
We need a screenshot like this one.
![Screenshot_20221117-112237_Settings](https://user-images.githubusercontent.com/18078393/202407984-a9efc3ef-c62c-454c-adaf-35f71bc2dd57.jpg)

**Comment by cristipaval** [comment]
I'm going to spend several mins today to try reproducing.

**Comment by cristipaval** [comment]
Boom! It crashed on my end as well, I was inspired to ask what keyboard you use, @adeel0202. After setting Gboard as the default keyboard on my Samsung, I am able to reproduce the crash. @JmillsExpensify, Gboard is the default keyboard on pure Android, I'll spend some time to investigate the issue.

**Comment by JmillsExpensify** [comment]
Thanks everyone for helping us get to the bottom of this one! Sounds great, let's keep it moving forward.

**Comment by cristipaval** [comment]
I investigated yesterday and today and seems that we are affected by [this](https://github.com/facebook/react-native/issues/30503#issuecomment-894825272) react issue. 
Here's why the cursor moves around wildly: `onSelectionChange` callback is fired twice: firstly with the position inside the emoji (between the high and low surrogates) and secondly the callback is fired with the position right before the emoji. The thing is that [we set the position on composer state](https://github.com/Expensify/App/blob/da8ffb2e002088c4686109114cb77705e625dd03/src/pages/home/report/ReportActionCompose.js#L197) when the callback is called and this creates infinite calls of the `onSelectionChange` callback alternating between the 2 cursor positions (inside and before the emoji.).
The crash happens when the cursor is "inside" emoji and the user hits delete button on keyboard. This deletes the high surrogate char of the emoji and the low surrogate remains as an invalid char which crashes the App.

@JmillsExpensify I think we should ask a RN expert to look at this issue and also check if my findings make sense.

**Comment by s77rt** [comment]
@adeel0202 I just tested with the proposed solution It didn't crash.
I didn't test prior the proposal, so maybe you should apply the proposal and test.

**Comment by JmillsExpensify** [comment]
For clarity, do we have a proposal that we're evaluating? If not, I'm going to assign the issue to one of our RN experts that help us with fixes upstream.

**Comment by s77rt** [comment]
@JmillsExpensify Please check on https://github.com/Expensify/App/issues/12854
I think this issue should be put on hold.
My proposal may fix this issue as well.

**Comment by JmillsExpensify** [comment]
Great point! I happened to land on one of the related issues shortly after the above message. We should consider holding all the linked issues in your Slack post on #12854, though I'll just do that for this issue for now given I'm assigned to it.

**Comment by JmillsExpensify** [comment]
Still on hold for linked issue.

**Comment by JmillsExpensify** [comment]
Still waiting on the linked issue, which is going to stake some time given that we're looking to submit a fix upstream in RN.

**Comment by JmillsExpensify** [comment]
Still working through #12854, though we're making progress there.

**Comment by JmillsExpensify** [comment]
Linked issue is currently held on an upstream fix.

**Comment by cristipaval** [comment]
You're annoying Melvin, this issue is on hold.

**Comment by JmillsExpensify** [comment]
Looks like the linked issue we're holding on is going to be unheld relatively soon.

**Comment by s77rt** [comment]
We are waiting for RN next release which could be in the next week

**Comment by JmillsExpensify** [comment]
That's super cool news! Thanks for the update.

**Comment by JmillsExpensify** [comment]
The latest RN release is out! Commented in the linked issue that we're still holding on.

**Comment by JmillsExpensify** [comment]
Looks like the latest estimation is that we'll get the fork updated this week.

**Comment by JmillsExpensify** [comment]
I think we can take this one off hold now that the fork is officially published, right?

**Comment by s77rt** [comment]
@JmillsExpensify This is on hold for https://github.com/Expensify/App/issues/12854 which is also on hold waiting for RN changes in our fork.

> now that the fork is officially published, right?

I think you are referring to [RNW](https://www.npmjs.com/package/@expensify/react-native-web). There is no recent update on [RN](https://www.npmjs.com/package/@expensify/react-native).

**Comment by JmillsExpensify** [comment]
Switching this to monthly given the urgency noted above.

**Comment by cristipaval** [comment]
Still on hold.

**Comment by JmillsExpensify** [comment]
Still on hold.

**Comment by cristipaval** [comment]
Still on hold.

**Comment by JmillsExpensify** [comment]
Alright, taking this issue off hold! That said, I think we should go ahead pay out the reporting bonus and close out this issue, creating a fresh one for the unknown symbol issue. Thoughts? @mananjadhav @cristipaval

**Comment by cristipaval** [comment]
Yes @JmillsExpensify, I do agree that we should create another issue for the unknown symbol. Please cc me on the newly created issue because I can remember some of the investigations I made a long time ago and I wanted to post my technical thoughts on it.

**Comment by mananjadhav** [comment]
I agree, we can create another issue for the unknown symbol. You can tag me if you want when we create the new issue.

**Comment by cristipaval** [comment]
Pinged @JmillsExpensify internally.

**Comment by JmillsExpensify** [comment]
@mananjadhav Will you be requesting payment via NewDot, or shall I issue via Upwork?

**Comment by mananjadhav** [comment]
@JmillsExpensify I don't think any payout for me here.

**Comment by cristipaval** [comment]
Only the reporting bonus should be paid to @adeel0202

**Comment by JmillsExpensify** [comment]
All paid out! I think we're done here. Closing.
